### PR TITLE
chore(deps): upgrade go to 1.21

### PR DESCRIPTION
**Pull request by janisz**
## Description

This PR bumps test containers to `0.3.66` that includes Go 1.21.8
In order to fix builds we need to run `go mod tidy` on all `go.mod` files that resulted in upgrading required version and added `toolchain`.
In `github.com/stackrox/rox/tests: TestEndpoints` the error has changed. In no client cert tests `expected cret` is probably a better error than `bar certificate` but I'm not sure why we get the same error when certificate is provided but is bad
To make prow test pass I replace go with 1.21.8 once prow image updated we can revert changes to lib.sh
- https://github.com/openshift/release/pull/49962
- https://github.com/stackrox/stackrox/pull/10490

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

CI
See replacement here https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/stackrox_stackrox/10305/pull-ci-stackrox-stackrox-master-shell-unit-tests/1771182198347534336/artifacts/shell-unit-tests/stackrox-initial/build-log.txt

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -470,7 +472,6 @@ replace (
 	github.com/sigstore/cosign/v2 => github.com/stackrox/cosign/v2 v2.0.0-20231206143943-626efb842da7
 
 	github.com/tecbot/gorocksdb => github.com/DataDog/gorocksdb v0.0.0-20200107201226-9722c3a2e063
-	go.uber.org/zap => github.com/stackrox/zap v1.15.1-0.20230918235618-2bd149903d0e
```
Comment (RTann) [inline go.mod:473]: do we no longer require the forked changes? were they merged upstream? just curious about why we can finally remove this

File: Makefile [hunk 2]
```diff
@@ -496,14 +496,14 @@ test-prep:
 .PHONY: go-unit-tests
 go-unit-tests: build-prep test-prep
 	set -o pipefail ; \
-	CGO_ENABLED=1 GODEBUG=cgocheck=2 MUTEX_WATCHDOG_TIMEOUT_SECS=30 GOTAGS=$(GOTAGS),test scripts/go-test.sh -timeout 15m -race -cover -coverprofile test-output/coverage.out -v \
+	CGO_ENABLED=1 GOEXPERIMENT=cgocheck2 MUTEX_WATCHDOG_TIMEOUT_SECS=30 GOTAGS=$(GOTAGS),test scripts/go-test.sh -timeout 15m -race -cover -coverprofile test-output/coverage.out -v \
```
Comment (RTann) [inline Makefile:499]: curious how come this changed

File: go.mod [hunk 3]
```diff
@@ -1,6 +1,8 @@
 module github.com/stackrox/rox
 
-go 1.19
+go 1.21
```
Comment (RTann) [inline go.mod:3]: do we do anything/depend on anything which requires 1.21?

File: Makefile (hunk 2 above)
Comment (janisz) [reply Makefile:499]: https://tip.golang.org/doc/go1.21#:~:text=GOEXPERIMENT%3Dcgocheck2

File: go.mod (hunk 3 above)
Comment (janisz) [reply go.mod:3]: `go mod tidy` adds it automatically :(

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,8 @@
...
+
+toolchain go1.21.7
```
[the remark this answers is not in the document]
Comment (janisz) [reply go.mod:5]: It's recomended version to compile the project and it's there by `go mod tidy` I don't know how to remove it :(

File: go.mod (hunk 1 above)
Comment (janisz) [reply go.mod:473]: I created a PR to investigate it https://issues.redhat.com/browse/ROX-23111
There are 2 features that we use (ordering of fields and not verbose errors output) 
I removed it here to see if it builds

Comment (janisz) [reply go.mod:473]:
```
go mod why -m go.uber.org/zap
# go.uber.org/zap
github.com/stackrox/rox/operator
go.uber.org/zap
```

Comment (RTann) [reply go.mod:473]: does this have to be done in this PR or can this be put into a separate one?

Comment (janisz) [reply go.mod:473]: I have a separated PR for that 
- https://github.com/stackrox/stackrox/pull/10375

**Comment by janisz** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10305?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#10305** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10305?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#10409** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10409?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10408** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10408?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10407** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10407?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10373** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10373?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#10390](https://github.com/stackrox/stackrox/pull/10390) <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10390?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#10375** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/10375?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @janisz and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: scripts/ci/lib.sh [hunk 5]
```diff
@@ -933,6 +933,24 @@ get_pr_details() {
...
+	GOLANG_SHA256=538b3b143dc7f32b093c8ffe0e050c260b57fc9d57a12c4140a639a8dd2b4e4f
+	url="https://dl.google.com/go/go$GOLANG_VERSION.linux-amd64.tar.gz"
+	wget --no-verbose -O go.tgz "$url"
+	echo "$GOLANG_SHA256 *go.tgz" | sha256sum -c -
+	ls
+	tar -C "$GOPATH" -xzf go.tgz
+	rm go.tgz
+	mkdir -p "$GOPATH/src" "$GOPATH/bin"
+	chmod -R 777 "$GOPATH"
+	go version
```
Comment (janisz) [inline scripts/ci/lib.sh:951]:
```suggestion
	wget --no-verbose -O /tmp/go.tgz "$url"
	echo "$GOLANG_SHA256 /tmp/go.tgz" | sha256sum -c -
	ls
	tar -C "$GOPATH" -xzf /tmp/go.tgz
	rm /tmp/go.tgz
	mkdir -p "$GOPATH/src" "$GOPATH/bin"
	chmod -R 777 "$GOPATH"
	go version
```

File: go.mod (hunk 4 above)
[the remark this answers is not in the document]
Comment (janisz) [reply go.mod:5]: On downstream we have 1.21.7 so keep it like this

File: scripts/ci/lib.sh [hunk 6]
```diff
@@ -933,6 +933,23 @@ get_pr_details() {
...
     info "BEGIN OpenShift CI mods"
 
+    #TODO(janisz): remove after https://github.com/openshift/release/pull/49962 is merged
+    if [ -e "/tmp/go/" ]; then
+        info "Go already exists"
+    else
+        info "Replace Go"
+        go version
+        whoami
+        GOLANG_VERSION=1.21.8
```
Comment (vikin91) [inline scripts/ci/lib.sh:943]: Shouldn't we use here 1.21.7 as in the other places in this PR (and in the downstream)?
```suggestion
        GOLANG_VERSION=1.21.8
```

**Review by vikin91 — APPROVED** [review]
Looks good!

File: scripts/ci/lib.sh (hunk 6 above)
Comment (janisz) [reply scripts/ci/lib.sh:943]: Yeah, I thought about that (and @RTann asked too) having difference in bugfix version shouldn't be a big deal. 
https://github.com/stackrox/rox-ci-image/pull/199#pullrequestreview-1932767312

**Review by rhybrillou — COMMENTED** [review]
Can you please also check the `.github/workflows/scanner-pr-vuln-update.yaml` file ?

**Review by rhybrillou — COMMENTED** [review]
Could you also check what is done in `tests/e2e/lib.sh` in the `customize_envVars` section around `GODEBUG` and `CGO_CHECKS` ?

**Comment by janisz** [comment]
@rhybrillou Thanks for pointing out. I've added label to run scanner update and fixed install via operator. I think it never worked as it only set `GODEBUG` value to `2` :shrug:

**Review by rhybrillou — APPROVED** [review]

**Pull request merged**